### ISSUE TITLE
Unnecessarily slow AbstractQuery.bind(int, Object)

### ISSUE THREAD

**Issue by bruto0** [jOOQ/jOOQ#7228]
### Expected behavior and actual behavior:
fast vs slow
### Steps to reproduce the problem:
this method in AbstractQuery creates an array on each invocation:

```java
public Query bind(int index, Object value) {
        Param<?>[] params = getParams().values().toArray(EMPTY_PARAM);
...
```

Now if your query contains _lots_ of parameters, this becomes very, very slow
Time these two snippets against each other and see:

```java
    private int numParams = 1000;
    Select<Record> createSelectQuery(int params) {
        SelectField[] fields = new Field[params];
        for (int i = 0;i < params;i++) {
            fields[i] = DSL.val(0L).as("param" + i);
        }
        return dsl.select(fields).from(table());
    }
    private ResultQuery query = createSelectQuery(numParams);

    @Test
    public void bindLoop() {
        ResultQuery<Record> resultQuery = query;
        for (int paramIdx = 1;paramIdx <= numParams;paramIdx++) {
            resultQuery.bind(paramIdx, paramIdx);
        }
    }

    @Test
    public void bindArray() {
        BatchBindStep batchBindStep = create.batch(query);
        Object[] paramValues = new Object[numParams];
        for (int paramIdx = 0;paramIdx < numParams;paramIdx++) {
            paramValues[paramIdx] = paramIdx;
        }
        batchBindStep.bind(paramValues);
    }
```

Unfortunately, BatchBindStep is of no use if you actually want your query to return something.

### Versions:
- jOOQ:3.10.5
- Java:1.8.0
- Database (include vendor): none, irrelevant to this issue
- OS: Windows 7
- JDBC Driver (include name if inofficial driver): n/a

**Comment by lukaseder** [comment]
Thanks for your report. Indeed, that particular logic is rather wasteful, but I think there are more profound improvements that could be made through #6616

I'll keep this in mind when implementing the other fix.

**Comment by lukaseder** [comment]
Instead of fixing the status quo, we can improve the API by making it immutable, which means that we no longer have to search for bind values in the expression tree. Closing this as won't fix, in favour of #11513
